pacta web build

maurolepore opened the pull request.

Commit:
tools to auto-build the pacta_web image and support files

Commit:
also pull the latest user_results repo

Commit:
set memeory to 4g

Commit:
get rid of stupid macOS files

Commit:
add wim-like script

Commit:
set files 666 and folders 777 in repo directories

Commit:
remove any previous 2dii_pacta image

Commit:
print the shas to stdout before deleting the clones

maurolepore committed:
Use for loops

maurolepore commented on pacta/cj_build_tag.sh:
```diff
@@ -0,0 +1,45 @@
+clones="PACTA_analysis create_interactive_report StressTestingModelDev pacta-data"
+url="git@github.com:2DegreesInvesting/"
+
+for repo in ${clones}
+do
+    remote="${url}${repo}.git"
+    git clone -b master ${remote} --depth 1
+    echo "--"
+done
```
@cjyetman, to more clearly see what I changed see 
https://github.com/2DegreesInvesting/2diidockerrunner/pull/13/commits/3a8e5d85762f0e3aaf70eae2e8014efc487c2012

cjyetman replied:
awesome! thanks!

maurolepore reviewed (commented):
@cjyetman I made this a PR so it's easier to work with this changes. The GH interface is nice to see who changed what and discuss it. 
Do you prefer to not merge this PR? (But even so the PR facilitates review)

cjyetman commented:
let's not merge this yet because I know for a fact there are substantial changes coming today

but thanks for opening the PR, I agree this is a bit better

maurolepore committed:
Extract 'tag'

maurolepore commented on pacta/cj_build_tag.sh:
```diff
@@ -0,0 +1,45 @@
+    echo "--"
+done
+
+docker rmi 2dii_pacta
```
This fails if there is more than one tag for the 2dii_pacta image:

```bash
➜  pacta git:(pacta-web-build) ✗ docker images                                
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
2dii_pacta          0.0.0.9000          43c79a68e414        3 minutes ago       3.36GB
2dii_pacta          1                   43c79a68e414        3 minutes ago       3.36GB
2dii_pacta          2                   43c79a68e414        3 minutes ago       3.36GB
maurolepore/pacta   0.0.1               358e8fe546f1        26 hours ago        3.71GB
2dii/r-packages     latest              2d1d6a7e7ade        44 hours ago        1.76GB
2dii/system         latest              9790f300c9cd        44 hours ago        1.26GB
ubuntu              18.04               56def654ec22        5 weeks ago         63.2MB

➜  pacta git:(pacta-web-build) ✗ docker rmi 2dii_pacta
Error: No such image: 2dii_pacta

➜  pacta git:(pacta-web-build) ✗ docker rmi 2dii_pacta:2
Untagged: 2dii_pacta:2

➜  pacta git:(pacta-web-build) ✗ docker rmi 2dii_pacta:1
Untagged: 2dii_pacta:1

➜  pacta git:(pacta-web-build) ✗ docker images          
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
2dii_pacta          0.0.0.9000          43c79a68e414        4 minutes ago       3.36GB
maurolepore/pacta   0.0.1               358e8fe546f1        26 hours ago        3.71GB
2dii/r-packages     latest              2d1d6a7e7ade        44 hours ago        1.76GB
2dii/system         latest              9790f300c9cd        44 hours ago        1.26GB
ubuntu              18.04               56def654ec22        5 weeks ago         63.2MB
```

cjyetman replied:
crap.... any ideas around that?

maurolepore replied:
I think it's no big deal -- a new tag overwrites the old one. We may either remove this line, leave it as is and just stare at the error message, or grep on "2dii_pacta*" and remove all images -- but that's too intrusive I think.

maurolepore replied:
Notice how I tag a new image with a preexisting tag, first from 2dii_pacta and later form maurolepore/pacta -- just to show how the image id changes depending on the source image:

```bash
➜  pacta git:(pacta-web-build) ✗ docker images                                
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
2dii_pacta          0.0.1.9000          90ad58717379        About an hour ago   3.36GB
2dii_pacta          1                   90ad58717379        About an hour ago   3.36GB
maurolepore/pacta   0.0.1               358e8fe546f1        28 hours ago        3.71GB
2dii/r-packages     latest              2d1d6a7e7ade        46 hours ago        1.76GB
2dii/system         latest              9790f300c9cd        46 hours ago        1.26GB
ubuntu              18.04               56def654ec22        5 weeks ago         63.2MB

➜  pacta git:(pacta-web-build) ✗ docker tag 2dii_pacta:1 2dii_pacta:0.0.1.9000 

➜  pacta git:(pacta-web-build) ✗ docker images                                
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
2dii_pacta          0.0.1.9000          90ad58717379        About an hour ago   3.36GB
2dii_pacta          1                   90ad58717379        About an hour ago   3.36GB
maurolepore/pacta   0.0.1               358e8fe546f1        28 hours ago        3.71GB
2dii/r-packages     latest              2d1d6a7e7ade        46 hours ago        1.76GB
2dii/system         latest              9790f300c9cd        46 hours ago        1.26GB
ubuntu              18.04               56def654ec22        5 weeks ago         63.2MB

➜  pacta git:(pacta-web-build) ✗ # Above, notice the number of tags 0.0.1.9000 (only one) and the image id compared to the tag 1

➜  pacta git:(pacta-web-build) ✗ docker tag maurolepore/pacta:0.0.1 2dii_pacta:0.0.1.9000                        

➜  pacta git:(pacta-web-build) ✗ docker images                                           
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
2dii_pacta          1                   90ad58717379        About an hour ago   3.36GB
2dii_pacta          0.0.1.9000          358e8fe546f1        28 hours ago        3.71GB
maurolepore/pacta   0.0.1               358e8fe546f1        28 hours ago        3.71GB
2dii/r-packages     latest              2d1d6a7e7ade        46 hours ago        1.76GB
2dii/system         latest              9790f300c9cd        46 hours ago        1.26GB
ubuntu              18.04               56def654ec22        5 weeks ago         63.2MB

➜  pacta git:(pacta-web-build) ✗ # Above, notice the number of tags 0.0.1.9000 (still one) and the image id compared to maurolepore/pacta


```

cjyetman replied:
versioning is definitely a better idea... like auto-incrementing on the highest tag/version that already exists?

what I did (`docker rmi 2dii_pacta`) was just the nuclear way of making sure I was loading/exporting the right thing

cjyetman replied:
we can get rid of this `docker rmi 2dii_pacta`... it is not really necessary, and it was meant to make things "safer", but it seems it's possible it makes things less safe/stable under certain conditions, so I think maybe just get rid of it. @maurolepore ?

maurolepore replied:
Yeah, I can't confirm right now but this line might relate to leaving a zombie image behind.

cjyetman replied:
eventually, I replaced this line with
`docker rmi --force $(docker images -q '2dii_pacta' | uniq)`
which searches for all images named "2dii_pacta" (regardless of tag), returns just the id/s of the image/s it finds, and `uniq`'s that list before sending it to `docker rmi`. This still causes a warning message if no images exist named "2dii_pacta", but the script is still able to continue as expected, so I think this is an adequate solution for now.

maurolepore committed:
Allow using as a command if added to PATH

maurolepore committed:
Allow piping if script makes it to the end

maurolepore committed:
Tag repos inside the container

If the container makes it to production, we can then spin the
container and push the tags to GitHub and call it a release.

Show where the tag points

maurolepore committed:
Use another for loop

maurolepore commented on pacta/cj_build_tag.sh:
```diff
@@ -0,0 +1,58 @@
+    echo "--"
+done
+
+for repo in ${clones}
+do
+    git -C "${repo}" tag -a "${tag}" -m "Release pacta ${tag}" HEAD
+    echo "${repo}"
+    echo "$(git -C ${repo} log --pretty='%h %d <%an> (%cr)' | head -n 1)"
+    echo "--"
+done
```
This now tags the temporary clone of each repo using the same tag as the container, then prints  a one-line log showing the newly added tag, the commit sha, message, commiter, and time. If the image makes it to production, we can push the tags from inside the final container with `git push origin <tagname>` -- which creates a GitHub release.

```bash
PACTA_analysis
15f92ff  (grafted, HEAD -> master, tag: 0.0.1.9000, origin/master, origin/HEAD) <Clare2D> (5 hours ago)
--
create_interactive_report
6234092  (grafted, HEAD -> master, tag: 0.0.1.9000, origin/master, origin/HEAD) <Clare2D> (84 minutes ago)
--
StressTestingModelDev
f89886d  (grafted, HEAD -> master, tag: 0.0.1.9000, tag: 0.0.1, origin/master, origin/HEAD) <jacobvjk> (6 days ago)
--
pacta-data
5e8ed2e  (grafted, HEAD -> master, tag: 0.0.1.9000, tag: 0.0.1, origin/master, origin/HEAD) <Clare2D> (20 hours ago)

```

cjyetman replied:
brilliant! thanks

maurolepore committed:
New pacta-clones

maurolepore committed:
Make pacta-clones executable

maurolepore committed:
WIP new pacta-push-tags

maurolepore committed:
Fix pacta-push-tags

maurolepore committed:
Default to pacta clones

cjyetman commented on pacta/cj_build_tag.sh:
```diff
@@ -0,0 +1,58 @@
+done
+
+docker rmi 2dii_pacta
+docker build ./ --tag 2dii_pacta:"${tag}"
```
this isn't playing well on my machine.... if I don't use an argument with the shell script command and there's no existing 2dii_pacta in my docker system, this loads a new image in my docker system with no name and spins up a zombie docker instance of that image which is a bit difficult to get rid of... so I'm currently removing the `:"${tag}"` bit of this on my local copy

maurolepore replied:
Oh — I’ll look into that

maurolepore replied:
I did a few experiments and I'm not able to reproduce the problem. This is surprising because I'm pretty sure I saw one of those `<none>` images recently. If you are able to reproduce, it'd be good to share screen or document more precisely how you got there.

cjyetman replied:
Now that I've used this numerous times in the past days and understand better the tagging thing, I don't see any real reason to use this script without setting the tag, so this problem I was having before is kind of irrelevant.

Commit:
create a .Rprofile to set appropriate latex settings for this non-user

Commit:
remove all tagged versions of 2dii_pacta

Commit:
tag as latest as well as version

Commit:
add pension funds example script and source files

Commit:
don't forget the dot

cjyetman commented:
Having used this a bunch of times now, and after numerous fixes and updates from @maurolepore and me, I think this is ready to merge.

cjyetman reviewed (approved):
I think this is ready to merge now

maurolepore committed:
Remove needless build-tag

maurolepore committed:
Remove needless pacta-clones

maurolepore committed:
Update readme

maurolepore committed:
Remove needless pacta-push-tags

maurolepore reviewed (commented):
@cjyetman, I pruned this PR  a bit:
* I removed commands we no longer use
* I updated README.

Can you review? I'm particularly interested in capturing your latest knowledge in README.

Commit:
Rename cj_build_tag.sh to build_with_tag.sh

Commit:
Update README.md

cjyetman commented:
I renamed the script to `build_with_tag.sh` and I had a go at the Readme. You should probably review it.

maurolepore reviewed (commented):
LGTM
Nit: Where it say ‘[commit] the tags’ I think it should say ‘push’.
